trieste/acquisition/rule.py

```diff
@@ -1103,6 +1104,324 @@ def state_func(
+
+            return state_, points
+
+        return state_func
+
+    def maybe_initialize_subspaces(
+        self,
+        subspaces: Sequence[UpdateableSearchSpaceType],
+        models: Mapping[Tag, ProbabilisticModelType],
+        datasets: Optional[Mapping[Tag, Dataset]] = None,
+    ) -> None:
+        """
+        Initialize subspaces if necessary.
+        Get a mask of subspaces that need to be initialized using an abstract method.
+        Initialize individual subpaces by calling the method of the UpdateableSearchSpaceType class.
+
+        This method can be overridden by subclasses to change this behaviour.
+        """
+        mask = self.get_initialize_subspaces_mask(subspaces, models, datasets)
+        tf.debugging.assert_equal(
+            tf.shape(mask),
+            (len(subspaces),),
+            message="The mask for initializing subspaces should be of the same length as the "
+            "number of subspaces",
+        )
+        for ix, subspace in enumerate(subspaces):
+            if mask[ix]:
+                subspace.initialize(models, datasets)
+
+    @abstractmethod
+    @check_shapes("return: [V]")
+    def get_initialize_subspaces_mask(
+        self,
+        subspaces: Sequence[UpdateableSearchSpaceType],
+        models: Mapping[Tag, ProbabilisticModelType],
+        datasets: Optional[Mapping[Tag, Dataset]] = None,
+    ) -> TensorType:
+        """
+        Return a boolean mask for subspaces that should be initialized.
+        This method is called during the acquisition step to determine which subspaces should be
+        initialized and which should be updated. The subspaces corresponding to True values in the
+        mask will be re-initialized.
+
+        :param subspaces: The sequence of subspaces.
+        :param models: The model for each tag.
+        :param datasets: The dataset for each tag.
+        :return: A boolean mask of length V, where V is the number of subspaces.
+        """
+        ...
+
+
+class TrustRegionBox(Box, UpdateableSearchSpace):
+    """An updateable box search space for use with trust region acquisition rules."""
+
+    def __init__(
+        self,
+        global_search_space: SearchSpace,
+        beta: float = 0.7,
+        kappa: float = 1e-4,
+        min_eps: float = 1e-2,
+    ):
+        """
+        Calculates the bounds of the box from the location/centre and global bounds.
+
+        :param global_search_space: The global search space this search space lives in.
+        """
+
+        self._global_search_space = global_search_space
+        self._beta = beta
+        self._kappa = kappa
+        self._min_eps = min_eps
+
+        super().__init__(global_search_space.lower, global_search_space.upper)
+
+    @property
+    def global_search_space(self) -> SearchSpace:
+        """The global search space this search space lives in."""
+        return self._global_search_space
+
+    @property
+    def location(self) -> TensorType:
```

Comment:
We generally don't use properties where plain attributes will do (like here and eps) as it's not considered Pythonic. mypy will still be able to typecheck assignments and usages based on the initialisation (and it's always possible to add properties later if we need to modify setting or getting).

Reply:
That makes sense for `location` and I'll remove it. However, we already have a use case for overriding `eps` (in #778), so I'll keep that one.

Reply:
Actually, as you have hinted, the derived class in #778 can add the `eps` properties there, so no need to keep them here :-)